drainer/sync/kafka.go

```diff
@@ -112,12 +142,60 @@ func (p *KafkaSyncer) Sync(item *Item) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+	switch p.partitionMode {
+	case PartitionFixed:
+		return errors.Trace(p.saveBinlog(slaveBinlog, item, nil))
+	case PartitionBySchema:
+		if slaveBinlog.Type == obinlog.BinlogType_DML {
+			return errors.Trace(p.syncDMLPartitionBySchema(slaveBinlog, item))
+		}
+		partitionKey := sarama.StringEncoder(*slaveBinlog.DdlData.SchemaName)
+		return errors.Trace(p.saveBinlog(slaveBinlog, item, partitionKey))
+	case PartitionByTable:
+		if slaveBinlog.Type == obinlog.BinlogType_DML {
+			return errors.Trace(p.syncDMLPartitionByTable(slaveBinlog, item))
+		}
+		partitionKey := sarama.StringEncoder(*slaveBinlog.DdlData.SchemaName + "." + *slaveBinlog.DdlData.TableName)
+		return errors.Trace(p.saveBinlog(slaveBinlog, item, partitionKey))
+	}
+	return errors.Errorf("Not supported partition mode: %v", p.partitionMode)
+}
 
-	err = p.saveBinlog(slaveBinlog, item)
-	if err != nil {
-		return errors.Trace(err)
+func (p *KafkaSyncer) syncDMLPartitionBySchema(binlog *obinlog.Binlog, item *Item) error {
+	tables := binlog.DmlData.Tables
+	if len(tables) == 1 {
+		return errors.Trace(p.saveBinlog(binlog, item, sarama.StringEncoder(*tables[0].SchemaName)))
+	}
+
+	tablesBySchema := make(map[string][]*obinlog.Table)
+	for _, table := range tables {
+		schemaName := *table.SchemaName
+		tablesBySchema[schemaName] = append(tablesBySchema[schemaName], table)
+	}
+
+	for schema, tables := range tablesBySchema {
+		binlog.DmlData.Tables = tables
+		if err := p.saveBinlog(binlog, item, sarama.StringEncoder(schema)); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (p *KafkaSyncer) syncDMLPartitionByTable(binlog *obinlog.Binlog, item *Item) error {
+	tables := binlog.DmlData.Tables
+	if len(tables) == 1 {
+		table := tables[0]
+		return errors.Trace(p.saveBinlog(binlog, item, sarama.StringEncoder(*table.SchemaName+"."+*table.TableName)))
```

Comment:
This specialization doesn't seem useful. The outcome of `len(tables) == 1` and `len(tables) != 1` are exactly the same.

Reply:
This can save one allocation. But it's small though and probably can be ignored.